RUM: Collecting browser errors (#9324)
DataDog/documentation: content/en/real_user_monitoring/browser/collecting_browser_errors.md

---
title: Collecting Browser Errors
kind: documentation
further_reading:
  - link: "/real_user_monitoring/error_tracking/"
    tag: Documentation
    text: "Error Tracking"
  - link: "https://www.datadoghq.com/blog/real-user-monitoring-with-datadog/"
    tag: "Blog"
    text: "Real User Monitoring"
  - link: "/real_user_monitoring/explorer/"
    tag: "Documentation"
    text: "Explore your views within Datadog"
  - link: "/real_user_monitoring/explorer/analytics/"
    tag: "Documentation"
    text: "Build analytics upon your events"
  - link: "/real_user_monitoring/dashboards/"
    tag: "Documentation"
    text: "RUM Dashboards"
---

Front-end errors are collected with Real User Monitoring (RUM). The error message and stack trace are included when available.

## Error Origins

> Front-end errors are split in 4 different categories depending on their `error.origin`:

Reviewer: ```suggestion
Front-end errors are split into four different categories depending on their `error.origin`:
```

> - **network**: XHR or Fetch errors resulting from AJAX requests. Specific attributes to network errors can be found [in the documentation][1].

Reviewer: ```suggestion
- **network**: XHR or Fetch errors resulting from AJAX requests. Specific attributes to network errors can be found [in the network errors documentation][1].
```

- **source**: Unhandled exceptions or unhandled promise rejections (source-code related).
- **console**: `console.error()` API calls.

> - **custom**: Errors sent with the [RUM `addError` API][2] default to `custom`.

Reviewer: ```suggestion
- **custom**: Errors sent with the [RUM `addError` API][2] that default to `custom`.
```

> ## Error attributes

Reviewer: ```suggestion
## Error Attributes
```

For information about the default attributes for all RUM event types, see [Data Collected][3]. For information about configuring for sampling, global context, or custom user actions, see [Advanced Configuration][4].

| Attribute       | Type   | Description                                                       |
|-----------------|--------|-------------------------------------------------------------------|
| `error.source`  | string | Where the error originates from (for example, `console` or `network`).     |
| `error.type`    | string | The error type (or error code in some cases).                   |
| `error.message` | string | A concise, human-readable, one-line message explaining the event. |
| `error.stack`   | string | The stack trace or complementary information about the error.     |

### Network errors

> Network errors include information about failing HTTP requests. The following facets are also collected:

Reviewer: ```suggestion
Network errors include information about failing HTTP requests. The following facets are collected:
```

| `error.resource.url_query` | object | The query string parts of the URL decomposed as query params key/value attributes. |
| `error.resource.url_scheme`      | string | The protocol name of the URL (HTTP or HTTPS).                                            |
| `error.resource.provider.name`      | string | The resource provider name. Default is `unknown`.                                            |
| `error.resource.provider.domain`      | string | The resource provider domain.                                            |
| `error.resource.provider.type`      | string | The resource provider type (for example `first-party`, `cdn`, `ad`, `analytics`).                                            |

### Source errors

Source errors include code-level information about the error. More information about the different error types can be found in [the MDN documentation][5].

| Attribute       | Type   | Description                                                       |
|-----------------|--------|-------------------------------------------------------------------|
| `error.type`    | string | The error type (or error code in some cases).                   |

> ## Collect errors manually

Reviewer: ```suggestion
## Collect Errors Manually
```

Monitor handled exceptions, handled promise rejections and other errors not tracked automatically by the RUM SDK with the `addError()` API:

> ```javascript

Reviewer: https://github.com/DataDog/dd-docs-tools/wiki/Markdown#code-blocks - ref the new code block as it has the c+p option.

addError(
    error: unknown,
    context?: Context,
    source: ErrorSource.CUSTOM | ErrorSource.NETWORK | ErrorSource.SOURCE = ErrorSource.CUSTOM
);
```

> **Note**: The [Error Tracking][6] feature processes errors sent with source set to `custom` or `source` and that contain a stacktrace.

Reviewer: ```suggestion
**Note**: The [Error Tracking][6] feature processes errors sent with source set to `custom` or `source` and that contain a stack trace.
```
